Image: Fix resizer controls being hidden in Safari when switching between alignments

andrewserong opened the pull request:
<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

<!-- Gutenberg's license is in the process of updating to be dual-licensed under the GPL and MPL. As part of that transition, all new contributions are dual-licensed. For more information, see: https://github.com/WordPress/gutenberg/blob/trunk/LICENSE.md -->

## Description
<!-- Please describe what you have changed or added -->

Fixes #24587 

In Safari, when switching the Image block between not aligned and an alignment, the component appears to re-render however the `img` element's `onload` callback is never fired, resulting in `naturalWidth` and `naturalHeight` being `undefined`. This means that the conditional [on this line](https://github.com/wordpress/gutenberg/blob/fffa06392bb584c1190948525a8767b78d9d9369/packages/block-library/src/image/image.js#L447) is reached, which means that the resizer isn't rendered.

The fix is to update `naturalWidth` and `naturalHeight` to pull from an image ref (so that the values are always available) and fall back to local component state when not available. This should 🤞 give us the best of both worlds — the real current value is available for Safari, and the fallback local component state value is available when cropping and editing images so that the saving state doesn't unexpectedly fail due to `undefined` natural width/height.

## How has this been tested?
<!-- Please describe in detail how you tested your changes. -->
<!-- Include details of your testing environment, tests ran to see how -->
<!-- your change affects other areas of the code, etc. -->

1. In Safari, before applying this PR, insert an image block.
2. Switch the image block from no alignment to any alignment (e.g. Left alignment)
3. Observe that the resizer controls are no longer visible.
4. Apply this PR.
5. Repeat the above steps, and the resizer should always be available.

Test in other browsers, and also test cropping / image editing to be sure this doesn't introduce any regressions.

## Screenshots <!-- if applicable -->

| Before | After |
| --- | --- |
| ![Kapture 2021-12-08 at 12 31 31](https://user-images.githubusercontent.com/14988353/145132409-b038f63b-b345-432c-9f51-c6eec1790770.gif) | ![Kapture 2021-12-08 at 12 29 20](https://user-images.githubusercontent.com/14988353/145132257-b61de180-e18f-4f90-a2c1-c420ba064218.gif) |

## Types of changes
<!-- What types of changes does your code introduce?  -->
<!-- Bug fix (non-breaking change which fixes an issue) -->
<!-- New feature (non-breaking change which adds functionality) -->
<!-- Breaking change (fix or feature that would cause existing functionality to not work as expected) -->

Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [ ] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [ ] I've tested my changes with keyboard and screen readers. <!-- Instructions: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/accessibility-testing.md -->
- [ ] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [ ] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/code/native-mobile.md -->

andrewserong committed:
Image: Update naturalWidth and naturalHeight to pull from image ref instead of local component state

andrewserong commented:
Update: this fix will need some adjustment, I just noticed it introduces a regression when dealing with external images when the external image is initially added to the block (the resizer is not visible).

We might wind up needing to retain the `onload` and use the imageRef as a fallback if available. I'll have a play around.

andrewserong committed:
Re-introduce onLoad setState behaviour to allow fallback while cropping images

andrewserong commented:
> I just noticed it introduces a regression when dealing with external images when the external image is initially added to the block (the resizer is not visible).

Another update: I _think_ I've fixed up the issue by preserving the local component state for the loaded image's natural width / height. We then use that as a fallback value, and prioritise the real `img` element's natural width/height when available, so that the fix for Safari still works.

paaljoachim commented:
Hey @andrewserong 
I assume we also need the backtrack label here as well?
As the various fixes can easier be picked up to be added to the next beta.

andrewserong commented:
> I assume we also need the backtrack label here as well?

@paaljoachim yes, we sure do! I've added the label. 👍

glendaviesnz commented:
This tests well for me. An alternative, rather than using image onload method at all for this, would be to use `ref.current.complete`. A quick test of just adding the following to the existing image.js code seems to work for chrome and safari:

```javascript
	const imgRef = useRef();
	const imageOnload = () => {
		setNaturalSize(
			pick( imgRef.current, [ 'naturalWidth', 'naturalHeight' ] )
		);
	};

	useEffect( () => {
		if ( imgRef.current?.complete ) {
			imageOnload();
		}
	}, [ imgRef.current?.complete ] );

	let img = (
		// Disable reason: Image itself is not meant to be interactive, but
		// should direct focus to block.
		/* eslint-disable jsx-a11y/no-noninteractive-element-interactions, jsx-a11y/click-events-have-key-events */
		<>
			<img
				ref={ imgRef }
				src={ temporaryURL || url }
				alt={ defaultedAlt }
				onError={ () => onImageError() }
			/>
			{ temporaryURL && <Spinner /> }
		</>
		/* eslint-enable jsx-a11y/no-noninteractive-element-interactions, jsx-a11y/click-events-have-key-events */
	);

```

glendaviesnz commented:
> A quick test of just adding the following to the existing image.js code seems to work for chrome and safari:

Actually it only works for existing images - `ref.current.complete` stays as false for new images for some reason.

glendaviesnz commented:
> Actually it only works for existing images - `ref.current.complete` stays as false for new images for some reason.

My bad, the useEffect in above should be:

```javascript
	useEffect( () => {
		if ( imgRef.current?.complete ) {
			imageOnload();
		}
	}, [ imgRef.current?.complete ] );
```
have updated the full example above, and now seems to test well for me for existing images on page load and new images added.

andrewserong commented:
Thanks for reviewing @glendaviesnz! Nice, I hadn't thought of using `imgRef?.current?.complete`. With the other approaches I tried for removing the `onload` handler, I wound up introducing bugs when using the cropping tool, or with external images that don't have CORS access. I'll give `imgRef?.current?.complete` a try! 🤞

Mamaduka commented on packages/block-library/src/image/image.js:
```diff
@@ -179,6 +183,23 @@ export default function Image( {
+				loadedNaturalHeight ||
+				undefined,
+		};
+	}, [ loadedNaturalWidth, loadedNaturalHeight, imageRef.current ] );
```
I don't think we need to include `imageRef.current` in the dependency, at least based on `exhaustive-deps` 😄

andrewserong replied:
Thanks! I've removed the `useMemo` call, since I _think_ setting the `naturalWidth` and `naturalHeight` isn't quite worth it (because there's no real calculation going on). Let me know if you think it's better to leave it in, and I can add it back 🙂

andrewserong replied:
Update: I actually just added it back in, as removing it caused the resizer to intermittently fail to display in Safari. I've included `imageRef.current?.complete` in the dependency array — here I _think_ we want the `useMemo` to recalculate _slightly_ more frequently than with just the `loadedNaturalWidth` and `loadedNaturalHeight` values, but happy to remove it if it winds up not being needed.

I've found the behaviour in Safari to be a bit unreliable in testing, so working out what is and isn't needed has been a bit more challenging than usual 😅

andrewserong committed:
Remove useMemo

andrewserong committed:
Re-instate useMemo as removing it caused the resizer to disappear intermittently

andrewserong commented:
Thanks for the suggestion to move away from the `onLoad` callback and to use `imageRef.current?.complete` in a `useEffect` instead, @glendaviesnz! Unfortunately doing that introduced a regression:

![Kapture 2021-12-10 at 17 09 47](https://user-images.githubusercontent.com/14988353/145525998-e9cf5cf7-82fc-403b-8c5c-015cb20d66ca.gif)

In this screengrab, I've added an external image, and once it's loaded, the resizer controls aren't available. I _think_ this is due to:

When inserting an image from an external URL, the `useEffect` [at this line](https://github.com/wordpress/gutenberg/blob/832d1c53e9bca886851050f26907a9a2e6c55339/packages/block-library/src/image/image.js#L169) attempts to async fetch the image so that it can be uploaded if need be. When the fetch fails for images without an appropriate CORS policy, the `imageRef.current?.complete` value is set to `false`, even though the image itself has loaded within the editor. This appears to be because fetching the asset resets `.complete`? And then the CORS failure means that it remains `false`. I found a tiny bit more context [in this MDN article](https://developer.mozilla.org/en-US/docs/Web/API/HTMLImageElement/complete), but I wasn't _too_ sure of what I was observing. It looks like the cases where the value changes are a fair bit more broad than the `onLoad` handler.

To resolve that issue, adding an `onLoad` handler back in on the `img` element worked, however we then have multiple state updates and additional renders occurring from having both the `useEffect` and `onLoad` methods updating the natural width/height. So, if we need to have the `onLoad` handler there, I think I prefer the idea of reducing the number of state updates and avoiding the `useEffect`? The `imageRef.current?.complete` is still a good idea, though, so I've added it to the `useMemo` dependency array. I've also moved things around a little bit to reduce the diff.

Let me know if you can think of a better way to handle all this, though, or if I missed something in your approach. Very happy to try alternatives!

glendaviesnz commented:
@andrewserong I didn't try the remote url scenario - I can't see a way around the CORs failure with my approach.  The current change of your's seems to work well in both cases - although I did have one failure still in Safari, so just trying to see if I can replicate that

glendaviesnz commented:
I couldn't replicate the issue with Safari not showing the resize handles with this patch applied, so not sure what caused that.

The only other thought I had was that maybe this extra code isn't needed if the `allowResize` context is set to false by a parent gallery, but I can't see an easy way of making it dependent on `allowResize=true` as hooks can't be used conditionally - but it doesn't seem like it would cause any performance issues on a large gallery, what do you think?

andrewserong commented:
Thanks for re-testing @glendaviesnz!

> The only other thought I had was that maybe this extra code isn't needed if the allowResize context is set to false by a parent gallery, but I can't see an easy way of making it dependent on allowResize=true as hooks can't be used conditionally - but it doesn't seem like it would cause any performance issues on a large gallery, what do you think?

That's a good point. I can't think of a clean way to exclude it, either, but 🤞 the `useMemo` means that the logic only fires when an image is loaded or changes are made to it, so I don't think it should have much of an effect on larger galleries. I just had a go at inserting a gallery with 50 images and performance switching between image blocks appears to be unaffected in my manual testing:

![Kapture 2021-12-13 at 10 46 24](https://user-images.githubusercontent.com/14988353/145734614-64e852c6-5838-4b28-b687-b67af936dc73.gif)

Let me know if you can think of any other edges cases I should test, though!

glendaviesnz commented:
> Let me know if you can think of any other edges cases I should test, though!

I can't think of any additional edge cases that we would need to check.

glendaviesnz reviewed (approved):
This tested well for me on chrome, firefox and safari.

ramonjd commented:
Also LGTM! 🙇
